File: TCP  IP Server Client Example Part 5 (Chat Application)/[C#]-TCP  IP Server Client Example Part 5 (Chat Application)/C#/ChatApp1.2/Forms/FrmServer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using NetComm;

namespace ChatApp1._2.Forms
{
    public partial class FrmServer : Form
    {
        private NetComm.Host server = new Host(3333); // Listens on port 3330.
        private int activeConnections = 0;


        private delegate void DisplayInformationDelegate(string s);

        private delegate void DisplayConnectionsDelegate(string s);

        private Dictionary<string, string> registeredUsers = new Dictionary<string, string>();
        private Dictionary<string, string> loggedInUsers = new Dictionary<string, string>();

        string startupPath = Application.StartupPath.ToString();

        public FrmServer()
        {
            InitializeComponent();
            startupPath = Path.Combine(startupPath, "users.db");
            LoadRegisteredUsers(); // for very large FlatFile databases this should be done on a separate thread. But really you should use a Database
            server.ConnectionClosed += Server_ConnectionClosed;
            server.DataReceived += Server_DataReceived;
            server.DataTransferred += Server_DataTransferred;
            server.errEncounter += ServerErrEncounter;
            server.lostConnection += ServerLostConnection;
            server.onConnection += ServerOnConnection;
            this.FormClosing += FrmServer_FormClosing;
        }

        void FrmServer_FormClosing(object sender, FormClosingEventArgs e)
        {
            SaveRegisteredUsers();
        }

        private void SaveRegisteredUsers()
        {
            // Not the best way to save the users.
            FileStream fs = null;
            StreamWriter sw = null;
            try
            {
                fs = new FileStream(startupPath, FileMode.OpenOrCreate, FileAccess.Write, FileShare.Inheritable);
                sw = new StreamWriter(fs);

                sw.WriteLine("# Updated Last on: " + DateTime.UtcNow.ToString());
                foreach (KeyValuePair<string, string> entry in registeredUsers)
                {
                    string line = entry.Key + "," + entry.Value; // longer appendings to string should be done through StringBuilder as it is so much faster!
                    sw.WriteLine(line);
                }

                sw.Close();
                fs.Close();
            }
            catch (Exception ex)
            {
                MessageBox.Show("SaveRegisteredUsers: Failed to load registered users" + Environment.NewLine + ex.Message);
                if (sw != null) sw.Close();
                if (fs != null) fs.Close();
            }
        }

        private void LoadRegisteredUsers()
        {
            FileStream fs = null;
            StreamReader sr = null;

            try
            {
                fs = new FileStream(startupPath, FileMode.OpenOrCreate, FileAccess.Read, FileShare.Inheritable);
                sr = new StreamReader(fs);
                while (!sr.EndOfStream) // While there is something to read.
                {
                    string line = sr.ReadLine();
                    if (line.StartsWith("#"))
                    {
                        // Ignore comments for just now
                    }
                    else if (string.IsNullOrWhiteSpace(line))
                    {
                        // ignore blank lines
                    }
                    else
                    {
                        // Process the line.
                        string[] parts = line.Split(','); // separate each value in the line into their respective parts.
                        // Part[0] is always the user's email address
                        // Part[1] is always the password for the user
                        try
                        {
                            registeredUsers.Add(parts[0].ToString(), parts[1].ToString());
                        }
                        catch (Exception ex)
                        {
                            // If we get here the database is probably corrupt!
                            MessageBox.Show("LoadRegisteredUsers: User's Database may be corrupt!" + Environment.NewLine + ex.Message);
                        }
                    }
                }
                sr.Close();
                fs.Close();
            }
            catch (Exception ex)
            {
                MessageBox.Show("LoadRegisteredUsers: Failed to load registered users" + Environment.NewLine + ex.Message);
                if (sr != null) sr.Close();
                if (fs != null) fs.Close();
            }
        }

        private void ServerOnConnection(string id)
        {
            lblConnections.Text = loggedInUsers.Count.ToString();
            DisplayInformation(String.Format("Client {0} Connected", id));
        }

        private void ServerLostConnection(string id)
        {
            lblConnections.Text = loggedInUsers.Count.ToString();
            DisplayInformation(String.Format("Client {0} Lost Connection", id));
        }

        private void ServerErrEncounter(Exception ex)
        {
            DisplayInformation("Server Error " + ex.Message);
        }

        private void Server_DataTransferred(string sender, string recipient, byte[] data)
        {

        }

        private void Server_DataReceived(string iD, byte[] data)
        {
            string message = ConvertBytesToString(data);
            if (iD == "0")
            {
                DisplayInformation("Received Command From Client " + iD + " for this Server");


                if (message.StartsWith("CLOSING"))
                {
                    loggedInUsers.Remove(iD);
                    DisplayConnection(loggedInUsers.Count.ToString());
                    DisplayInformation("Client " + iD + " is Closing");
                }
                else if (message.StartsWith("REGISTRATION"))
                {
                    // Message = REGISTRATION:<USERNAME>:<PASSWORD>
                    if (registeredUsers.ContainsKey(message.Split(':')[1].ToString()))
                    {
                        server.SendData(iD, ConvertStringToBytes("ERROR:That email is already registered."));
                        DisplayInformation("Client " + iD + "ERROR:That email is already registered.");
                    }
                    else
                    {
                        registeredUsers.Add(message.Split(':')[1].ToString(), message.Split(':')[2].ToString());
                        loggedInUsers.Add(iD, DateTime.Now.ToString());
                        DisplayConnection(loggedInUsers.Count.ToString());
                        server.SendData(iD, ConvertStringToBytes("SUCCESS:Logged On"));
                        DisplayInformation("Client " + iD + "SUCCESS registered.");
                    }
                }
                else if (message.StartsWith("LOGIN"))
                {
                    // Message = LOGIN:<USERNAME>:<PASSWORD>
                    string[] parts = message.Split(':');
                    if (registeredUsers.ContainsKey(parts[1].ToString()))
                    {
                        string res = "";
                        if (registeredUsers.TryGetValue(parts[1].ToString(), out res))
                        {
                            if (res == parts[2])
                            {
                                // Successfully logged in
                                int newSenderId = (loggedInUsers.Count + 1);
                                server.SendData(iD, ConvertStringToBytes("ID:" + newSenderId.ToString()));
                                loggedInUsers.Add(newSenderId.ToString(), DateTime.Now.ToString());
                                DisplayConnection(loggedInUsers.Count.ToString());
                                DisplayInformation("Client " + parts[1].ToString() + " SUCCESS Logged in.");

                                
                            }
                            else
                            {
                                server.SendData(iD, ConvertStringToBytes("ERROR:Login Details Not Recognized"));
                                DisplayInformation("Client " + iD + "ERROR:Login Details Not Recognized");
                            }
                        }
                    }
                    else
                    {
                        server.SendData(iD, ConvertStringToBytes("ERROR:Login Details Not Recognized"));
                        DisplayInformation("Client " + iD + "ERROR:Login Details Not Recognized");
                    }
                }
                else
                {
                    DisplayInformation("Received Command From Client " + iD + " for this Server: " + message);
                }


            }
            else
            {
                DisplayInformation("Received Command From Client " + iD + " for this Server: " + message);
            }
        }

        private void Server_ConnectionClosed()
        {
            DisplayInformation("Connection Was Closed");
        }

        private void NewToolStripMenuItemClick(object sender, EventArgs e)
        {
            server.StartConnection();
            lblConnectionStatusImage.Image = Properties.Resources.green_circle_md;
        }

        void DisplayConnection(string s)
        {
            if (this.toolStrip1.InvokeRequired)
            {
                DisplayConnectionsDelegate d = new DisplayConnectionsDelegate(DisplayConnection);
                this.toolStrip1.Invoke(d, new object[] { s });
            }
            else
            {
                lblConnections.Text = s;
            }

        }


        void DisplayInformation(string s)
        {
            if (this.rtbConOut.InvokeRequired)
            {
                DisplayInformationDelegate d = new DisplayInformationDelegate(DisplayInformation);
                this.rtbConOut.Invoke(d, new object[] { s });
            }
            else
            {
                this.rtbConOut.AppendText(Environment.NewLine + s);
                this.rtbConOut.Focus();
            }
        }

        string ConvertBytesToString(byte[] bytes)
        {
            return ASCIIEncoding.ASCII.GetString(bytes);
        }

        byte[] ConvertStringToBytes(string str)
        {
            return ASCIIEncoding.ASCII.GetBytes(str);
        }

        private void OpenToolStripMenuItemClick(object sender, EventArgs e)
        {
            server.CloseConnection();
            lblConnectionStatusImage.Image = Properties.Resources.red_47690_640;
        }

        private void BtnNewClient_Click(object sender, EventArgs e)
        {
            FrmChatMain chat = new FrmChatMain();
            chat.Show();
        }
    }
}

File: TCP  IP Server Client Example Part 5 (Chat Application)/[VB]-TCP  IP Server Client Example Part 5 (Chat Application)/VB/ChatApp1.2/Forms/FrmServer.vb
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports NetComm

Namespace ChatApp1._2.Forms
	Partial Public Class FrmServer
		Inherits Form

		Private server As NetComm.Host = New Host(3333) ' Listens on port 3330.
		Private activeConnections As Integer = 0


		Private Delegate Sub DisplayInformationDelegate(ByVal s As String)

		Private Delegate Sub DisplayConnectionsDelegate(ByVal s As String)

		Private registeredUsers As New Dictionary(Of String, String)()
		Private loggedInUsers As New Dictionary(Of String, String)()

		Private startupPath As String = Application.StartupPath.ToString()

		Public Sub New()
			InitializeComponent()
			startupPath = Path.Combine(startupPath, "users.db")
			LoadRegisteredUsers() ' for very large FlatFile databases this should be done on a separate thread. But really you should use a Database
			AddHandler server.ConnectionClosed, AddressOf Server_ConnectionClosed
			AddHandler server.DataReceived, AddressOf Server_DataReceived
			AddHandler server.DataTransferred, AddressOf Server_DataTransferred
			AddHandler server.errEncounter, AddressOf ServerErrEncounter
			AddHandler server.lostConnection, AddressOf ServerLostConnection
			AddHandler server.onConnection, AddressOf ServerOnConnection
			AddHandler Me.FormClosing, AddressOf FrmServer_FormClosing
		End Sub

		Private Sub FrmServer_FormClosing(ByVal sender As Object, ByVal e As FormClosingEventArgs)
			SaveRegisteredUsers()
		End Sub

		Private Sub SaveRegisteredUsers()
			' Not the best way to save the users.
			Dim fs As FileStream = Nothing
			Dim sw As StreamWriter = Nothing
			Try
				fs = New FileStream(startupPath, FileMode.OpenOrCreate, FileAccess.Write, FileShare.Inheritable)
				sw = New StreamWriter(fs)

				sw.WriteLine("# Updated Last on: " & Date.UtcNow.ToString())
				For Each entry As KeyValuePair(Of String, String) In registeredUsers
					Dim line As String = entry.Key & "," & entry.Value ' longer appendings to string should be done through StringBuilder as it is so much faster!
					sw.WriteLine(line)
				Next entry

				sw.Close()
				fs.Close()
			Catch ex As Exception
				MessageBox.Show("SaveRegisteredUsers: Failed to load registered users" & Environment.NewLine & ex.Message)
				If sw IsNot Nothing Then
					sw.Close()
				End If
				If fs IsNot Nothing Then
					fs.Close()
				End If
			End Try
		End Sub

		Private Sub LoadRegisteredUsers()
			Dim fs As FileStream = Nothing
			Dim sr As StreamReader = Nothing

			Try
				fs = New FileStream(startupPath, FileMode.OpenOrCreate, FileAccess.Read, FileShare.Inheritable)
				sr = New StreamReader(fs)
				Do While Not sr.EndOfStream ' While there is something to read.
					Dim line As String = sr.ReadLine()
					If line.StartsWith("#") Then
						' Ignore comments for just now
					ElseIf String.IsNullOrWhiteSpace(line) Then
						' ignore blank lines
					Else
						' Process the line.
						Dim parts() As String = line.Split(","c) ' separate each value in the line into their respective parts.
						' Part[0] is always the user's email address
						' Part[1] is always the password for the user
						Try
							registeredUsers.Add(parts(0).ToString(), parts(1).ToString())
						Catch ex As Exception
							' If we get here the database is probably corrupt!
							MessageBox.Show("LoadRegisteredUsers: User's Database may be corrupt!" & Environment.NewLine & ex.Message)
						End Try
					End If
				Loop
				sr.Close()
				fs.Close()
			Catch ex As Exception
				MessageBox.Show("LoadRegisteredUsers: Failed to load registered users" & Environment.NewLine & ex.Message)
				If sr IsNot Nothing Then
					sr.Close()
				End If
				If fs IsNot Nothing Then
					fs.Close()
				End If
			End Try
		End Sub

		Private Sub ServerOnConnection(ByVal id As String)
			lblConnections.Text = loggedInUsers.Count.ToString()
			DisplayInformation(String.Format("Client {0} Connected", id))
		End Sub

		Private Sub ServerLostConnection(ByVal id As String)
			lblConnections.Text = loggedInUsers.Count.ToString()
			DisplayInformation(String.Format("Client {0} Lost Connection", id))
		End Sub

		Private Sub ServerErrEncounter(ByVal ex As Exception)
			DisplayInformation("Server Error " & ex.Message)
		End Sub

		Private Sub Server_DataTransferred(ByVal sender As String, ByVal recipient As String, ByVal data() As Byte)

		End Sub

		Private Sub Server_DataReceived(ByVal iD As String, ByVal data() As Byte)
			Dim message As String = ConvertBytesToString(data)
			If iD = "0" Then
				DisplayInformation("Received Command From Client " & iD & " for this Server")


				If message.StartsWith("CLOSING") Then
					loggedInUsers.Remove(iD)
					DisplayConnection(loggedInUsers.Count.ToString())
					DisplayInformation("Client " & iD & " is Closing")
				ElseIf message.StartsWith("REGISTRATION") Then
					' Message = REGISTRATION:<USERNAME>:<PASSWORD>
					If registeredUsers.ContainsKey(message.Split(":"c)(1).ToString()) Then
						server.SendData(iD, ConvertStringToBytes("ERROR:That email is already registered."))
						DisplayInformation("Client " & iD & "ERROR:That email is already registered.")
					Else
						registeredUsers.Add(message.Split(":"c)(1).ToString(), message.Split(":"c)(2).ToString())
						loggedInUsers.Add(iD, Date.Now.ToString())
						DisplayConnection(loggedInUsers.Count.ToString())
						server.SendData(iD, ConvertStringToBytes("SUCCESS:Logged On"))
						DisplayInformation("Client " & iD & "SUCCESS registered.")
					End If
				ElseIf message.StartsWith("LOGIN") Then
					' Message = LOGIN:<USERNAME>:<PASSWORD>
					Dim parts() As String = message.Split(":"c)
					If registeredUsers.ContainsKey(parts(1).ToString()) Then
						Dim res As String = ""
						If registeredUsers.TryGetValue(parts(1).ToString(), res) Then
							If res = parts(2) Then
								' Successfully logged in
								Dim newSenderId As Integer = (loggedInUsers.Count + 1)
								server.SendData(iD, ConvertStringToBytes("ID:" & newSenderId.ToString()))
								loggedInUsers.Add(newSenderId.ToString(), Date.Now.ToString())
								DisplayConnection(loggedInUsers.Count.ToString())
								DisplayInformation("Client " & parts(1).ToString() & " SUCCESS Logged in.")


							Else
								server.SendData(iD, ConvertStringToBytes("ERROR:Login Details Not Recognized"))
								DisplayInformation("Client " & iD & "ERROR:Login Details Not Recognized")
							End If
						End If
					Else
						server.SendData(iD, ConvertStringToBytes("ERROR:Login Details Not Recognized"))
						DisplayInformation("Client " & iD & "ERROR:Login Details Not Recognized")
					End If
				Else
					DisplayInformation("Received Command From Client " & iD & " for this Server: " & message)
				End If


			Else
				DisplayInformation("Received Command From Client " & iD & " for this Server: " & message)
			End If
		End Sub

		Private Sub Server_ConnectionClosed()
			DisplayInformation("Connection Was Closed")
		End Sub

		Private Sub NewToolStripMenuItemClick(ByVal sender As Object, ByVal e As EventArgs) Handles newToolStripMenuItem.Click
			server.StartConnection()
			lblConnectionStatusImage.Image = My.Resources.green_circle_md
		End Sub

		Private Sub DisplayConnection(ByVal s As String)
			If Me.toolStrip1.InvokeRequired Then
				Dim d As New DisplayConnectionsDelegate(AddressOf DisplayConnection)
				Me.toolStrip1.Invoke(d, New Object() { s })
			Else
				lblConnections.Text = s
			End If

		End Sub


		Private Sub DisplayInformation(ByVal s As String)
			If Me.rtbConOut.InvokeRequired Then
				Dim d As New DisplayInformationDelegate(AddressOf DisplayInformation)
				Me.rtbConOut.Invoke(d, New Object() { s })
			Else
				Me.rtbConOut.AppendText(Environment.NewLine & s)
				Me.rtbConOut.Focus()
			End If
		End Sub

		Private Function ConvertBytesToString(ByVal bytes() As Byte) As String
			Return ASCIIEncoding.ASCII.GetString(bytes)
		End Function

		Private Function ConvertStringToBytes(ByVal str As String) As Byte()
			Return ASCIIEncoding.ASCII.GetBytes(str)
		End Function

		Private Sub OpenToolStripMenuItemClick(ByVal sender As Object, ByVal e As EventArgs) Handles openToolStripMenuItem.Click
			server.CloseConnection()
			lblConnectionStatusImage.Image = My.Resources.red_47690_640
		End Sub

		Private Sub BtnNewClient_Click(ByVal sender As Object, ByVal e As EventArgs) Handles BtnNewClient.Click
			Dim chat As New FrmChatMain()
			chat.Show()
		End Sub
	End Class
End Namespace
